SPIFBlockDevice.cpp

```diff
@@ -17,368 +17,1019 @@
+    bool erase_failed = false;
+    int status = SPIF_BD_ERROR_OK;
+    // Find region of erased address
+    int region = _utils_find_addr_region(addr);
+    // Erase Types of selected region
+    uint8_t bitfield = _region_erase_types_bitfield[region];
 
-    for (uint32_t i = 0; i < addrc; i++) {
-        _spi.write(0xff & (addr >> 8*(addrc-1 - i)));
+    tr_debug("DEBUG: erase - addr: %llu, in_size: %llu", addr, in_size);
+
+    // For each iteration erase the largest section supported by current region
+    while (size > 0) {
+
+        // iterate to find next Largest erase type ( a. supported by region, b. smaller than size)
+        // find the matching instruction and erase size chunk for that type.
+        type = _utils_iterate_next_largest_erase_type(bitfield, size, (unsigned int)addr, _region_high_boundary[region]);
+        cur_erase_inst = _erase_type_inst_arr[type];
+        chunk = _erase_type_size_arr[type];
+
+        tr_debug("DEBUG: erase - addr: %llu, size:%d, Inst: 0x%xh, chunk: %d , ",
+                 addr, size, cur_erase_inst, chunk);
+        tr_debug("DEBUG: erase - Region: %d, Type:%d",
+                 region, type);
+
+        _mutex->lock();
+
+        if (_set_write_enable() != 0) {
+            tr_error("ERROR: SPI Erase Device not ready - failed");
+            erase_failed = true;
+            status = SPIF_BD_ERROR_READY_FAILED;
+            goto exit_point;
+        }
+
+        _spi_send_erase_command(cur_erase_inst, addr, size);
+
+        addr += chunk;
+        size -= chunk;
+
+        if ( (size > 0) && (addr > _region_high_boundary[region]) ) {
+            // erase crossed to next region
+            region++;
+            bitfield = _region_erase_types_bitfield[region];
+        }
+
+        if ( false == _is_mem_ready()) {
+            tr_error("ERROR: SPI After Erase Device not ready - failed\n");
+            erase_failed = true;
+            status = SPIF_BD_ERROR_READY_FAILED;
+            goto exit_point;
+        }
+
+        _mutex->unlock();
     }
 
-    for (uint32_t i = 0; i < retc; i++) {
-        rets[i] = _spi.write(0);
+exit_point:
+    if (erase_failed) {
+        _mutex->unlock();
     }
-    _cs = 1;
 
-    if (SPIF_DEBUG) {
-        printf("spif <- %02x", op);
-        for (uint32_t i = 0; i < addrc; i++) {
-            if (i < addrc) {
-                printf("%02lx", 0xff & (addr >> 8*(addrc-1 - i)));
-            } else {
-                printf("  ");
+    return status;
+}
+
+bd_size_t SPIFBlockDevice::get_read_size() const
+{
+    // Assuming all devices support 1byte read granularity
+    return SPIF_DEFAULT_READ_SIZE;
+}
+
+bd_size_t SPIFBlockDevice::get_program_size() const
+{
+    // Assuming all devices support 1byte program granularity
+    return SPIF_DEFAULT_PROG_SIZE;
+}
+
+bd_size_t SPIFBlockDevice::get_erase_size() const
+{
+    // return minimal erase size supported by all regions (0 if none exists)
+    return _min_common_erase_size;
```

Comment:
Maybe this should assert if there's no common erase size? Most of the existing code doesn't expect errors from this function.

Reply:
SFDP's JEDEC Standard specifies several examples where memory is divided into several sector regions - some of them support only 4KB erase, and some only 64KB erase - I think we shouldn't assert at such case:
![image](https://user-images.githubusercontent.com/36957743/44307209-e2d04a00-a3a6-11e8-8378-343d24912db6.png)

Reply:
@geky Maybe this seems harsh, but shouldn't we deprecate this function (the one that doesn't accept the address as parameter). We have this problem in other cases as well (FlashIAP for instance).

Reply:
What does LittleFS call then? LittleFS only support fixed-size blocks, but that's the majority of the storage devices (especially larger ones).

My thought was if the blockdevice-level get_erase_size gets called on a device without a common block size then it's an error (shouldn't use a block-based filesystem on a block device without a common block size).

I double checked with LittleFS and FAT and it looks like both will error if get_erase_size returns 0, so maybe returning zero is a good idea.
- [FAT](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/fat/ChaN/ff.cpp#L3277)
- [LittleFS](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/littlefs/lfs.c#L2073-L2079)